### ISSUE TITLE
Discover jumbo frame support

### ISSUE THREAD

**Issue by awh** [weaveworks/weave#1877]
End users must currently specify a large `WEAVE_MTU` to take advantage of jumbo frames on compliant hardware; this is required to achieve the highest possible fast datapath throughput on high-speed (e.g. 10G+) networks. Weave should automatically discover the optimum value for this variable if possible.

**Comment by dpw** [comment]
Any idea for how to do this? I'm not sure there is a simple notion of optimum value, and even to the extent that there is, how it can be reliably determined.

I expect that the default value of WEAVE_MTU is fine for the vast majority of weave net users.  The main reason to tweak it is for benchmarking games.  Attempts at heuristics to choose a larger MTU under certain conditions might well disadvantage real users more than it benefits them.

**Comment by awh** [comment]
> Any idea for how to do this?

No.

> Attempts at heuristics to choose a larger MTU under certain conditions might well disadvantage real users more than it benefits them.

I agree (see discussion in #net yesterday) - will icebox for now.

**Comment by rade** [comment]
Why would the discovery process disadvantage real users? Especially if we have do not have "any idea how to do this" yet.

**Comment by awh** [comment]
> Why would the discovery process disadvantage real users? Especially if we do not have "any idea how to do this" yet.

It's a hunch - I think you would have to experimentally raise the MTU of the network interfaces, which  might cause traffic loss? Needs more thought.

**Comment by rade** [comment]
I wasn't suggesting that weave modify the MTU of "physical" interfaces. Rather it should a) set the WEAVE_MTU based on the MTU of the physical interface(s) (which interfaces? good question), and b) check that this is working and either fall back to a safe MTU (i.e. same as current default) or probe for a better MTU.

**Comment by dpw** [comment]
> Why would the discovery process disadvantage real users? Especially if we do not have "any idea how to do this" yet.

For most real users, the benefits of increasing WEAVE_MTU when that is safe are less significant than the penalty of falling back to sleeve due to choosing a WEAVE_MTU that is too high.  So if in doubt, it's better to stick with the current default for WEAVE_MTU.  I'm sceptical that there is any way to automatically determine that increasing WEAVE_MTU is safe with sufficient confidence, for any set of circumstances.  So it's always better to stick with the low WEAVE_MTU.

**Comment by dpw** [comment]
> set the WEAVE_MTU based on the MTU of the physical interface(s) (which interfaces? good question)

But it's the PMTU between hosts that matters, not the MTU set on the interfaces.  Those are only the same in simple cases.  And the set of hosts isn't fixed.

**Comment by rade** [comment]
Good point. Having a single MTU has its limitations. And I guess proper PMTU discovery - where it feeds all the way to the containers - is hard / impossible with the OVS/vxlan kernel support as it stands, and/or hampered by network kit dropping jumbo frames w/o sending ICMPs.

**Comment by gvenka008c** [comment]
We are running Linux Kernel version 3.10.0. Does weave strictly enforces to have versions 3.12 in order to support fast data path?

```
 The required open vSwitch datapath (ODP) and VXLAN features are present in Linux kernel versions 3.12 and greater. If your kernel was built without the necessary modules Weave Net will fall back to the "user mode" packet path.
```

++ How do we verify the underlying network delivers packets of the size specified plus overheads of around 50 bytes?

**Comment by awh** [comment]
Hi @gvenka008c 

> Does weave strictly enforces to have versions 3.12 in order to support fast data path?

Yes - earlier Linux kernels simply don't have the features required for fast datapath to work I'm afraid.

> How do we verify the underlying network delivers packets of the size specified plus overheads of around 50 bytes?

You can use the `-s` option to ping to try different packet sizes; make sure you enable the DF (don't fragment) bit with `-M do`

**Comment by gvenka008c** [comment]
@awh Thanks. So when we run in sleeve mode, does it defaults to 1410 bytes (MTU)? We are running CentOS 3.10.0-327.28.2.el7.x86_64. We disabled Encryption and Opened UDP port 6784. We didn't specify and default value for WEAVE_MTU. When we restart weave, it shows fastdatapath but immediately switches to sleeve mode. Thoughts?

```
# weave version
weave script 1.7.2
weave router 1.7.2
weave proxy  1.7.2
weave plugin 1.7.2
```

**Comment by bboreham** [comment]
@gvenka008c please could you open a new issue rather than tagging on to this one which is about a specific enhancement.
